power_grid_model_io_native_c/power_grid_model_io_native/include/power_grid_model_io_native/pgm_vnf_converter/pgm_vnf_converter.hpp

```diff
@@ -83,123 +79,53 @@ inline PgmVnfConverter::PgmVnfConverter(std::string_view buffer, ExperimentalFea
-        std::cout << "CreationTime: " << match[2].str() << "\n";
-        std::cout << "Name: " << match[3].str() << "\n";
-        std::cout << "Unom: " << match[4].str() << "\n\n";
-        double const unom = std::stod(match[4].str());
-
-        all_nodes.emplace_back(node_id, unom);
-
-        // all_nodes["ID"].push_back(std::to_string(node_id));
-        // all_nodes["Unom"].push_back(match[4].str());
-
-        ++node_id;
-        ++it;
-    }
+    auto parser = PgmVnfParser(this->buffer_);
+    parser.parse_input();
+    this->parsed_vnf_data_ = parser.get_parsed_data();
+    this->id_lookup_ = parser.get_id_lookup();
+}
 
-    this->parsed_nodes_ = all_nodes;
+inline pgm::ConstDataset PgmVnfConverter::make_const_dataset(pgm::meta_data::MetaData const& meta_data) {
+    pgm::ConstDataset const_dataset{false, 1, "input", meta_data};
+    const_dataset.add_buffer("node", nodes_.size(), nodes_.size(), nullptr, nodes_.data());
+    return const_dataset;
 }
 
 inline void PgmVnfConverter::convert_input() {
-    std::vector<pgm::NodeInput> const nodes = convert_node_input();
-    convert_line_input();
-    convert_sources_input();
-    convert_sym_loads_input();
-    convert_shunts_input();
-    convert_transformer_input();
-    convert_sym_gens_input();
-    convert_links_input();
-
-    // pgm::Container<pgm::Node> container;
-
-    // for (auto const& node : nodes) {
-    //     container.emplace<pgm::Node>(node.id(), node);
-    // }
+    convert_node_input();
 
+    // construction complete has to be after we convert all the components
+    this->deserialized_data_.set_construction_complete();
     constexpr auto const& meta_data = pgm::meta_data::meta_data_gen::meta_data;
-    pgm::ConstDataset const const_dataset = make_const_dataset(nodes, meta_data);
-
+    pgm::ConstDataset const const_dataset = make_const_dataset(meta_data);
     std::string const serialized_pgm_data = serialize_data(const_dataset);
-
-    // 1. our vnf importer it directly understands the vnf format
-    // 2. convert vnf like dataset to internal types (take raw data and convert it to pgm component container)
-    // 3. another function which 1. makes const dataset view from pgm component container  2. const dataset to
-    // serializer
-
     this->serialized_data_ = serialized_pgm_data;
 }
 
 inline void PgmVnfConverter::set_file_buffer(std::string_view file_buffer) { this->buffer_ = file_buffer; }
 
-inline void PgmVnfConverter::set_deserialized_dataset(pgm::WritableDataset* data) { this->deserialized_data_ = data; }
+inline void PgmVnfConverter::set_deserialized_dataset(PgmInput data) { this->deserialized_data_ = std::move(data); }
 
 inline std::string_view PgmVnfConverter::get_file_buffer() const { return this->buffer_; }
 
-inline pgm::WritableDataset* PgmVnfConverter::get_deserialized_dataset() { return this->deserialized_data_; }
+inline PgmInput PgmVnfConverter::get_deserialized_dataset() const { return this->deserialized_data_; }
```

Comment:
maybe good to make this a struct because some formats may require both `PgmInput` and `PgmUpdate`, e.g.

```cpp
struct PgmConvertedData {
    PgmInput input;
    PgmUpdate update;
};
```
alternatively you can directly use `using ConvertedData = PgmInput` in the class definition